src/config/ConfigBuilder.ts

```diff
@@ -91,15 +95,85 @@ export class ConfigBuilder {
             } else if (key === 'customLogger') {
                 this.handleLogger(value);
             } else if (key === 'customCredentials') {
-                this.effectiveConfig.customCredentials = jsonObject;
+                this.effectiveConfig.customCredentials = value;
             } else if (key === 'backupAckToClientEnabled') {
                 this.effectiveConfig.backupAckToClientEnabled = tryGetBoolean(value);
+            } else if (key === 'security') {
+                this.handleSecurity(value);
             } else {
                 throw new RangeError(`Unexpected config key '${key}' is passed to the Hazelcast Client`);
             }
         }
     }
 
+    private static validateSecurityConfiguration(jsonObject: any): void {
+        if ('security' in jsonObject && 'customCredentials' in jsonObject) {
+            throw new RangeError('Ambiguous security configuration is found. ' +
+                'Use one of \'security\' or \'customCredentials\' elements, not both.')
+        }
+    }
+
+    private handleSecurity(jsonObject: any): void {
+        let isCredentialsSet = false;
+        for (const key in jsonObject) {
+            if (isCredentialsSet) {
+                throw new RangeError('Security configuration may only contain one of the supported credential types. ' +
+                    'Multiple credential types are passed to the Hazelcast Client.');
+            }
+
+            const value = jsonObject[key];
+            if (key === 'usernamePassword') {
+                this.handleUsernamePasswordCredentials(value);
+            } else if (key === 'token') {
+                this.handleTokenCredentials(value);
+            } else if (key === 'custom') {
+                this.effectiveConfig.security.custom = value;
+            } else {
+                throw new RangeError(`Unexpected security config ${key} is passed to the Hazelcast Client`);
+            }
+
+            isCredentialsSet = true;
+        }
+    }
+
+    private handleUsernamePasswordCredentials(jsonObject: any): void {
+        let username: string;
+        let password: string;
+        for (const key in jsonObject) {
+            const value = jsonObject[key];
+            if (key === 'username') {
+                username = tryGetStringOrNull(value);
+            } else if (key === 'password') {
+                password = tryGetStringOrNull(value);
+            } else {
+                throw new RangeError(`Unexpected username password credentials option '${key}' is passed to the Hazelcast Client`)
+            }
+        }
+
+        this.effectiveConfig.security.usernamePassword = new UsernamePasswordCredentialsImpl(username, password);
```

Comment:
We can have username and/or password `undefined` as a result of this, but it must be a string or null. Isn't this the case? We can fix that by validating both username and password are provided. Another way to fix this is that initializing username and password variables with `null` in `handleUsernamePasswordCredentials`

Actually `CodecUtil.encodeNullable` encodes undefined as `NULL_FRAME` since it checks for `value == null`:

```ts
static encodeNullable<T>(clientMessage: ClientMessage, value: T, encoder: (msg: ClientMessage, val: T) => void): void {
        if (value == null) {
            clientMessage.addFrame(NULL_FRAME.copy());
        } else {
            encoder(clientMessage, value);
        }
    }
```

Still, I think it is better to prevent `undefined` value to avoid confusing some users or hazelcast developers while debugging.

Reply:
I think it would be best to initialize them with null :+1: